app/copy.py

```diff
@@ -0,0 +1,214 @@
+
+    # apply weapon
+    arthur["power"] += arthur["weapon"]["power"]
+
+    # apply potion if exist
+    if arthur["potion"] is not None:
+        if "power" in arthur["potion"]["effect"]:
+            arthur["power"] += arthur["potion"]["effect"]["power"]
+
+        if "protection" in arthur["potion"]["effect"]:
+            arthur["protection"] += arthur["potion"]["effect"]["protection"]
+
+        if "hp" in arthur["potion"]["effect"]:
+            arthur["hp"] += arthur["potion"]["effect"]["hp"]
+
+    # mordred
+    mordred = knightsConfig["mordred"]
+
+    # apply armour
+    mordred["protection"] = 0
+    for a in mordred["armour"]:
+        mordred["protection"] += a["protection"]
+
+    # apply weapon
+    mordred["power"] += mordred["weapon"]["power"]
+
+    # apply potion if exist
+    if mordred["potion"] is not None:
+        if "power" in mordred["potion"]["effect"]:
+            mordred["power"] += mordred["potion"]["effect"]["power"]
+
+        if "protection" in mordred["potion"]["effect"]:
+            mordred["protection"] += mordred["potion"]["effect"]["protection"]
+
+        if "hp" in mordred["potion"]["effect"]:
+            mordred["hp"] += mordred["potion"]["effect"]["hp"]
+
+    # red_knight
+    red_knight = knightsConfig["red_knight"]
+
+    # apply armour
+    red_knight["protection"] = 0
+    for a in red_knight["armour"]:
+        red_knight["protection"] += a["protection"]
+
+    # apply weapon
+    red_knight["power"] += red_knight["weapon"]["power"]
+
+    # apply potion if exist
+    if red_knight["potion"] is not None:
+        if "power" in red_knight["potion"]["effect"]:
+            red_knight["power"] += red_knight["potion"]["effect"]["power"]
+
+        if "protection" in red_knight["potion"]["effect"]:
+            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
+
+        if "hp" in red_knight["potion"]["effect"]:
+            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+
+    # -------------------------------------------------------------------------------
+    # BATTLE:
+
+    # 1 Lancelot vs Mordred:
+    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
+    mordred["hp"] -= lancelot["power"] - mordred["protection"]
+
+    # check if someone fell in battle
+    if lancelot["hp"] <= 0:
+        lancelot["hp"] = 0
+
+    if mordred["hp"] <= 0:
+        mordred["hp"] = 0
+
+    # 2 Arthur vs Red Knight:
+    arthur["hp"] -= red_knight["power"] - arthur["protection"]
+    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
+
+    # check if someone fell in battle
+    if arthur["hp"] <= 0:
+        arthur["hp"] = 0
+
+    if red_knight["hp"] <= 0:
+        red_knight["hp"] = 0
+
+    # Return battle results:
+    return {
+        lancelot["name"]: lancelot["hp"],
+        arthur["name"]: arthur["hp"],
+        mordred["name"]: mordred["hp"],
+        red_knight["name"]: red_knight["hp"],
+    }
+
+
+print(battle(KNIGHTS))
```

Comment:
The code for preparing each knight and simulating battles is repetitive. Consider creating helper functions to apply armour, weapon, and potion effects to reduce redundancy. This will make the code more maintainable and easier to read.